proxy/server.go

```diff
@@ -18,74 +18,168 @@
 package proxy
 
 import (
-	"bufio"
-	"fmt"
+	"io"
+	"log"
 	"net"
 	"net/http"
+	"net/http/httputil"
 	"strings"
 
-	"github.com/elazarl/goproxy"
+	"github.com/inconshreveable/go-vhost"
+	"github.com/pkg/errors"
+	"github.com/soheilhy/cmux"
 	netproxy "golang.org/x/net/proxy"
 )
 
-type stickyHasher interface {
-	Hash(ip string) string
-}
-
 type domainTracker interface {
 	Inc(domain string)
 }
 
+type proxyServer struct {
+	dialer  netproxy.Dialer
+	sm      StickyMapper
+	dt      domainTracker
+	portMap map[string]string
+}
+
+type StickyMapper interface {
+	Save(ip, userID string)
+	Hash(ip string) (hash string)
+}
+
 // NewServer returns new instance of HTTP transparent proxy server
-func NewServer(addr string, upstreamDialer netproxy.Dialer, mapper stickyHasher, dt domainTracker) *http.Server {
-	server := goproxy.NewProxyHttpServer()
-	server.Verbose = true
-	server.NonproxyHandler = http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		if req.Host == "" {
-			fmt.Fprintln(w, "Cannot handle requests without Host header, e.g., HTTP 1.0")
-			return
-		}
-		req.URL.Scheme = "http"
-		req.URL.Host = req.Host
-		server.ServeHTTP(w, req)
-	})
-
-	server.ConnectDial = upstreamDialer.Dial
-	server.OnRequest().HandleConnectFunc(func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
-		return goproxy.OkConnect, host
-	})
-
-	server.OnRequest().DoFunc(func(req *http.Request, ctx *goproxy.ProxyCtx) (*http.Request, *http.Response) {
-		conn, err := upstreamDialer.Dial("tcp", req.Host+":80")
-		if err != nil {
-			return req, nil
-		}
+func NewServer(upstreamDialer netproxy.Dialer, mapper StickyMapper, dt domainTracker, portMap map[string]string) *proxyServer {
+	return &proxyServer{
+		dialer:  upstreamDialer,
+		sm:      mapper,
+		dt:      dt,
+		portMap: portMap,
+	}
+}
+
+func (s *proxyServer) ListenAndServe(addr string) {
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.Fatalf("Error listening for https connections - %v", err)
```

Comment:
It's same as panicing in the middle of code, let's bubble errors

Reply:
Done.